### PR TITLE
perf(moe): fuse FlashInfer NVFP4 routing-map padding initialization

### PR DESCRIPTION
**Pull request by byshiue**
## Summary

Move routing-map padding initialization into the existing FlashInfer TRT-LLM NVFP4 routing kernels, eliminating the separate full-map `cudaMemsetAsync` which is introduced by PR https://github.com/lightseekorg/tokenspeed/pull/1569 and mentioned by https://github.com/lightseekorg/tokenspeed/pull/1569#discussion_r4046315346. We will check that if this issue should be fixed by flashinfer directly later.

The routing map must mark unused entries as invalid so the following GEMM cannot read stale token indices. This change preserves that guarantee while avoiding an extra initialization pass:

- Each expert’s final tile initializes its padded rows.
- Routing threads initialize unused allocation capacity and any guard entry.
- Valid assignments and padding have disjoint writers.
- Initialization runs on every invocation, including CUDA graph replay, and preserves programmatic dependent launch (PDL) ordering.

The adapter passes the map’s actual allocated length through the routing parameters. It builds matching sources and headers in a source-keyed private JIT module, with compatibility checks that reject unrecognized upstream layouts.

This applies to both routing from logits and precomputed top-k assignments.

## Performance

Measured with the real 93-layer Kimi-K3 NVFP4 model, TP8 on eight GB300 GPUs in the same NVLink domain, using FlashInfer 0.6.18. Outer, KDA, and decode CUDA graphs and overlap were enabled in both versions. Timing runs used no NSYS.

Comparison: main at `ffa16b13` versus this change at `3d408d33`.

### Agentic workflow replay

The benchmark replays frozen, tokenized SWE-smith conversations. Each run contains four conversations × eleven turns: 44 requests, with 500 generated tokens per request.

Results are median wall times from two independently restarted server processes per version, run in before–after–after–before order. Timings include measured request processing and client handoffs; model loading, capture, and warmup are excluded.

| Concurrent conversations | Before | After | E2E time reduction |
| --- | ---: | ---: | ---: |
| 1 | 235.039 s | 220.972 s | 5.99% |
| 2 | 137.452 s | 133.642 s | 2.77% |
| 4 | 84.730 s | 81.089 s | 4.30% |

The clearest benefit is in decode. At concurrency 1, median request-visible decode time per token decreased from **9.950 ms to 9.311 ms**, a **6.42% reduction**.

Fixed incremental-prefill measurements remained close to parity: latency changes ranged from **−1.13% to +0.88%** across 15 input/batch configurations covering BS 1–4, with 12 measured batches per configuration and version.

Concurrency denotes active conversations, not a fixed forward batch size. Concurrent runs can form different batches and cache hits; output differences also occurred between same-version repetitions. The serial comparison had identical outputs and cache hits in both repetitions. With two process repetitions, these results are measured improvements rather than statistical guarantees.

## Correctness and validation

- **66 regression tests passed**, covering routing from logits and precomputed top-k, PDL on/off, local expert partitions, empty local work, and multiple routing producers.
- CUDA graph tests deliberately corrupt the routing workspace between replays and verify that valid mappings and invalid padding are refreshed correctly.
- Real-model checks covered **30 paired batches / 70 paired requests** under matched cold and warm cache conditions. Inputs, cache-hit counts, and all eight greedy output tokens matched before and after.
- All **360 measured fixed-prefill batches** retained before/after output parity.
- The E2E benchmark completed **528 requests / 264,000 generated tokens**, with no request failures or preemptions.
- `pre-commit run --all-files` passed.

### CODE REVIEW

File: tokenspeed-kernel/python/tokenspeed_kernel/thirdparty/flashinfer/_routing_padding.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+The native source patterns come from FlashInfer's NVIDIA TRT-LLM routing code
+(https://github.com/flashinfer-ai/flashinfer), licensed under Apache-2.0. See
+the FlashInfer entry in the distributed THIRDPARTYNOTICES for source paths,
+upstream copyrights and license terms. The MIT header above covers TokenSpeed's
+original adapter and padding helpers, not the upstream code being transformed.
+"""
+
+import re
+
+_HELPERS = r"""
```
[the remark this answers is not in the document]
Comment (byshiue) [reply tokenspeed-kernel/python/tokenspeed_kernel/thirdparty/flashinfer/_routing_padding.py:35]: I am not sure is it a bug of flashinfer or not.
If this is a bug of flashinfer, we will file a PR to fix that on flashinfer and revert the change on this part.

[the remark this answers is not in the document]
Comment (byshiue) [reply tokenspeed-kernel/python/tokenspeed_kernel/thirdparty/flashinfer/_routing_padding.py:35]: I have filed a issue in flashinfer https://github.com/flashinfer-ai/flashinfer/issues/5349 

If we hope to have better performance in short term, we can merge this PR first, and update the codes later. Otherwise, let's wait the reply of flashinfer and decide how to fix this issue.

[the remark this answers is not in the document]
Comment (byshiue) [reply tokenspeed-kernel/python/tokenspeed_kernel/thirdparty/flashinfer/_routing_padding.py:35]: I have discussed with flashinfer team and we think this issue should be fixed in flashinfer side directly. 
I will file a PR in flashinfer.
